File: stress/metrics/gds.py
from __future__ import annotations

from dataclasses import dataclass
from typing import List, Optional

from OCRB.measure.events import Event, EventType


@dataclass(frozen=True)
class GDSResult:
    gds: Optional[float]               # None = N/A
    n_levels: int
    stress_levels: List[float]
    completion_rates: List[float]
    na_reason: Optional[str] = None


def compute_gds(
    events: List[Event],
    expected_levels: Optional[List[float]] = None,
) -> GDSResult:
    """
    BP-1 — Graceful Degradation Score (GDS)

    Spec-aligned definition:
      Execute at ordered stress intensity levels s1..sn
      Measure completion rate Ci at each level
      GDS = (1/n) * Σ Ci

    Data source:
      We accept per-level completion evidence as events with:
        - type == WORK_UNIT_END (optional per-task) OR
        - events that directly carry (stress_level, completion_rate)
      For v0 reference implementation, we prefer the explicit evidence events:
        Event with completion_rate != None and stress_level != None.

    expected_levels:
      If provided, enforce that we have data for each declared level (orderable).
    """
    levels: List[float] = []
    rates: List[float] = []

    # Prefer explicit evidence events (stress_level + completion_rate)
    for e in events:
        if e.stress_level is not None and e.completion_rate is not None:
            levels.append(float(e.stress_level))
            rates.append(float(e.completion_rate))

    if not levels:
        return GDSResult(
            gds=None,
            n_levels=0,
            stress_levels=[],
            completion_rates=[],
            na_reason="No (stress_level, completion_rate) evidence found in events.",
        )

    # Basic bounds check
    for r in rates:
        if not (0.0 <= r <= 1.0):
            return GDSResult(
                gds=None,
                n_levels=len(levels),
                stress_levels=levels,
                completion_rates=rates,
                na_reason=f"completion_rate out of bounds: {r}",
            )

    # If expected levels are declared, enforce coverage
    if expected_levels is not None:
        exp = [float(x) for x in expected_levels]
        got = set(levels)
        missing = [x for x in exp if x not in got]
        if missing:
            return GDSResult(
                gds=None,
                n_levels=len(levels),
                stress_levels=levels,
                completion_rates=rates,
                na_reason=f"Missing declared stress levels: {missing}",
            )

    n = len(rates)
    gds = sum(rates) / n
    gds = max(0.0, min(1.0, gds))

    # Ensure stress levels are monotonically orderable (spec requirement)
    # We don’t force monotonic in events ordering, but we ensure they can be sorted and disclosed.
    paired = sorted(zip(levels, rates), key=lambda x: x[0])
    s_sorted = [p[0] for p in paired]
    c_sorted = [p[1] for p in paired]

    return GDSResult(
        gds=gds,
        n_levels=n,
        stress_levels=s_sorted,
        completion_rates=c_sorted,
        na_reason=None,
    )

File: stress-ref/src/metrics/gds.rs
use crate::types::events::Event;

#[derive(Debug, Clone)]
pub struct GdsResult {
    pub gds: Option<f64>,
    pub n_levels: usize,
    pub stress_levels: Vec<f64>,
    pub completion_rates: Vec<f64>,
    pub na_reason: Option<String>,
}

/// BP-1: Graceful Degradation Score.
/// GDS = (1/n) * sum(C_i) where C_i is completion rate at stress level i.
pub fn compute_gds(events: &[Event], expected_levels: Option<&[f64]>) -> GdsResult {
    let mut levels = Vec::new();
    let mut rates = Vec::new();

    for e in events {
        if let (Some(sl), Some(cr)) = (e.stress_level, e.completion_rate) {
            levels.push(sl);
            rates.push(cr);
        }
    }

    if levels.is_empty() {
        return GdsResult {
            gds: None, n_levels: 0,
            stress_levels: vec![], completion_rates: vec![],
            na_reason: Some("No (stress_level, completion_rate) evidence found".to_string()),
        };
    }

    for &r in &rates {
        if !(0.0..=1.0).contains(&r) {
            return GdsResult {
                gds: None, n_levels: levels.len(),
                stress_levels: levels, completion_rates: rates,
                na_reason: Some(format!("completion_rate out of bounds: {r}")),
            };
        }
    }

    // Tolerance-based expected level enforcement
    if let Some(expected) = expected_levels {
        let missing: Vec<f64> = expected
            .iter()
            .filter(|&&exp| !levels.iter().any(|&got| (exp - got).abs() < 1e-9))
            .copied()
            .collect();
        if !missing.is_empty() {
            return GdsResult {
                gds: None, n_levels: levels.len(),
                stress_levels: levels, completion_rates: rates,
                na_reason: Some(format!("Missing declared stress levels: {missing:?}")),
            };
        }
    }

    let n = rates.len() as f64;
    let gds = (rates.iter().sum::<f64>() / n).clamp(0.0, 1.0);

    // Sort by stress level for disclosure
    let mut paired: Vec<(f64, f64)> = levels.into_iter().zip(rates).collect();
    paired.sort_by(|a, b| a.0.partial_cmp(&b.0).unwrap_or(std::cmp::Ordering::Equal));

    GdsResult {
        gds: Some(gds),
        n_levels: paired.len(),
        stress_levels: paired.iter().map(|p| p.0).collect(),
        completion_rates: paired.iter().map(|p| p.1).collect(),
        na_reason: None,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::types::events::{EventLog, EventType};

    #[test]
    fn perfect_gds() {
        let mut log = EventLog::new("t", "W1-A");
        log.emit(EventType::RunStart);
        for &level in &[0.1, 0.2, 0.3] {
            let e = log.emit(EventType::WorkUnitEnd);
            e.stress_level = Some(level);
            e.completion_rate = Some(1.0);
        }
        log.emit(EventType::RunEnd);

        let result = compute_gds(log.events(), Some(&[0.1, 0.2, 0.3]));
        assert_eq!(result.gds, Some(1.0));
    }

    #[test]
    fn degrading_gds() {
        let mut log = EventLog::new("t", "W1-A");
        log.emit(EventType::RunStart);
        for (&level, &rate) in [0.1, 0.2, 0.3].iter().zip(&[1.0, 0.5, 0.0]) {
            let e = log.emit(EventType::WorkUnitEnd);
            e.stress_level = Some(level);
            e.completion_rate = Some(rate);
        }
        log.emit(EventType::RunEnd);

        let result = compute_gds(log.events(), Some(&[0.1, 0.2, 0.3]));
        assert!((result.gds.unwrap() - 0.5).abs() < 0.01);
    }

    #[test]
    fn no_evidence_na() {
        let mut log = EventLog::new("t", "W1-A");
        log.emit(EventType::RunStart);
        log.emit(EventType::RunEnd);
        let result = compute_gds(log.events(), None);
        assert!(result.gds.is_none());
    }
}
